src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentation.java

```diff
@@ -84,34 +82,32 @@ private void setAggressivelyBatching(boolean aggressivelyBatching) {
             this.aggressivelyBatching = aggressivelyBatching;
         }
 
-        private void enterList() {
-            synchronized (this) {
-                stack.push(true);
-            }
+        private void setOutstandingFieldFetches(int level, int count) {
+            outstandingFieldFetchCounts.put(level, count);
         }
 
-        private void exitList() {
-            synchronized (this) {
-                if (!stack.isEmpty()) {
-                    stack.pop();
-                }
-            }
+        // TODO: should be threadsafe
```

Comment:
To make this threadsafe - invent a new type that encapsulates level + count and make that have read / write locks.